db/schema_applier.cc

```diff
@@ -259,22 +261,32 @@ static future<std::set<sstring>> merge_keyspaces(distributed<service::storage_pr
      *   to this node
      */
     auto diff = difference(before, after, indirect_equal_to<lw_shared_ptr<query::result_set>>());
+    auto sk_diff = difference(sk_before, sk_after, indirect_equal_to<lw_shared_ptr<query::result_set>>());
 
     auto& created = diff.entries_only_on_right;
     auto& altered = diff.entries_differing;
     auto& dropped = diff.entries_only_on_left;
 
+    altered.insert(sk_diff.entries_differing.begin(), sk_diff.entries_differing.end());
+    for (auto&& ks : boost::range::join(sk_diff.entries_only_on_right, sk_diff.entries_only_on_left)) {
```

Comment:
what are you trying to achieve with this logic? I don't understand it.
1. better naming is needed, use aliases like above "auto& dropped", it's hard to deduce what is "left" or "right".
2. also sk_after is not a good name, it's impossible to deduce what sk stands for here, without looking at upper calls.
3. I feel that even if this logic somehow works it should not be coded like this, especially why it goes against diffing mechanism from above? it seems that for the code to be generic you need to apply the diff in all cases not just alteration

Reply:
For the motivation of this logic, please see https://github.com/scylladb/scylladb/pull/20852#discussion_r1790013589

Reply:
this comment doesn't explain motivation or how it's supposed to work, it only contains code snippet.

I think this code is difficult to read, we should change it in a way that it can be understood without reading gh issues or PR comments.

I'd suggest commenting what it does, it's not trivial, fix variable names they are quite bad and contribute to unreadability.

Reply:
I assume you only meant to refactor the newly introduced source code.
The names are somewhat adjusted to the naming convention used here. For example `sk_diff` stands for "system_schema.scylla_keyspaces diff", where (within this function) by "diff" we understand the difference between the keyspace from before and after schema modification, e.g. executing a CQL statement. The variable `diff`, that was here already, stores the diff from system_schema.keyspaces, which is a basis for all the logic in this function, so `sk_diff` just extends this name. As this name sits with me well, I'm unsure how would you like to change it, I'm opened for a sugestion (and I don't think using the full name is the best option).

The code snippet that this specific comment is attached to does 1 thing: it pre-fills the `altered` set, which later results in performing the side effects of executing ALTER. Now, when considering the differences of a keyspace coming from 2 system tables, that is: system_schema.keyspaces and system_schema.scylla_keyspaces, we must cover 2 extra scenarios
1. a row in system_schema.scylla_keyspaces, describing a ks, has changed so that non-null columns in this row have different values now, in which case the ks is being altered, in which case we need to insert these modifications into `altered`, which is covered by the first line from this snippet,
2. the same as above but for a case where a column is nulled or was null and now has a value, which requires alter being performed, but this alter is only required if a ks isn't being created (because this schema modification is going to be taken care of by `created`) nor dropped (because you cannot modify a ks during a drop statement).

I can also suggest reader the banner comment at the top of this function, maybe it's going to help.

In any case, please do let me know how you want to progress here.

Reply:
> I can also suggest reader the banner comment at the top of this function, maybe it's going to help.

you didn't add any comment but added logic how old comments are supposed to explain new logic?

> we must cover 2 extra scenarios 

that should be explained in the code

Reply:
I added comments, please do let me know if it's enough or I should elaborate more.